### PR TITLE
sharness: Generate JUnit test reports

### PR DESCRIPTION
**Pull request by magik6k**
This makes sharness generate a JUnit XML report that can be used by Jenkins to view test results in a bit nicer form than a huge txt file (removing the fun of wasting few minutes to find the failure each time one happens). The report is saved as `test/sharness/test-results/sharness.xml`.

Example report: https://ipfs.io/ipfs/QmZVCjX2L9DrbqYaVeVzTXZZUbyQvTM3mTUmzeMCoN8Pqc/sharness.xml - [HTML from junit2html](https://ipfs.io/ipfs/QmZVCjX2L9DrbqYaVeVzTXZZUbyQvTM3mTUmzeMCoN8Pqc/sharness.xml.html) - Jenkins should show failures at the top ([like here](https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fjs-ipfs-unixfs-engine/detail/PR-200/2/tests))

Somewhat depends on https://github.com/ipfs/jenkins/pull/85

### CODE REVIEW

**Comment by whyrusleeping** [comment]
> (removing the fun of wasting few minutes to find the failure each time one happens)

Aww, but that was my favorite part

**Comment by whyrusleeping** [comment]
@magik6k @VictorBjelkholm whats the blocker here?

**Comment by magik6k** [comment]
I'd like to have test result collection enabled on jenkins before this is merged to make sure it works well with our setup (though I have checked it with local jenkins instance and it did work).

That's done in https://github.com/ipfs/jenkins/pull/85, though @VictorBjelkholm suggests to defer/integrate those changes to https://github.com/ipfs/go-ipfs/pull/4430 / https://github.com/ipfs/jenkins/issues/81 (Which will probably take more time, but it needs to be done either way)

TL;DR stuff in this PR Works For Me(tm), I'd like to see it not breaking production Jenkins before it's merged

**Comment by magik6k** [comment]
tbh I'd love to have this enabled already, even with the legacy job as it doesn't appear the new one is getting enabled anytime soon, and looking through failed sharness output is getting annoying.

I rebased this, shouldn't break anything, some review would be welcome.

File: test/sharness/Rules.mk [hunk 1]
```diff
@@ -29,12 +29,17 @@ export MAKE_SKIP_PATH=1
...
+	-@(cd $(@D) && ./$(@F)) 2>&1
+else
 	@(cd $(@D) && ./$(@F)) 2>&1
+endif
 .PHONY: $(T_$(d))
 
 $(d)/aggregate: $(T_$(d))
 	@echo "*** $@ ***"
 	@(cd $(@D) && ./lib/test-aggregate-results.sh)
+	@(cd $(@D) && ./lib/gen-junit-report.sh)
```
Comment (Kubuxu) [inline test/sharness/Rules.mk]: I would make this separate target that corresponds to created file.

File: test/sharness/Rules.mk [hunk 2]
```diff
@@ -29,14 +29,23 @@ export MAKE_SKIP_PATH=1
...
 	@(cd $(@D) && ./$(@F)) 2>&1
+endif
 .PHONY: $(T_$(d))
 
 $(d)/aggregate: $(T_$(d))
 	@echo "*** $@ ***"
 	@(cd $(@D) && ./lib/test-aggregate-results.sh)
 .PHONY: $(d)/aggregate
 
+$(d)/junit_xml: $(T_$(d))
```
Comment (Kubuxu) [inline test/sharness/Rules.mk]: I meant, name it `test-results/sharness.xml` as it produces this file.

File: test/sharness/Rules.mk [hunk 3]
```diff
@@ -52,7 +61,7 @@ $(d)/deps: $(SHARNESS_$(d)) $$(DEPS_$(d)) # use second expansion so coverage can
 test_sharness_deps: $(d)/deps
 .PHONY: test_sharness_deps
 
-test_sharness_short: $(d)/aggregate
+test_sharness_short: $(d)/aggregate $(d)/junit_xml
```
Comment (Kubuxu) [inline test/sharness/Rules.mk]: Most users won't need the JUnit xmls so there is no reason to create them for them.

File: test/sharness/lib/0001-Generate-partial-JUnit-reports.patch [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
++}
++
+ test -n "$test_description" || error "Test script did not set test_description."
+ 
+ if test "$help" = "t"; then
+@@ -251,30 +261,69 @@ test_have_prereq() {
+ 	test $total_prereq = $ok_prereq
+ }
+ 
++junit_testcase() {
```
Comment (Kubuxu) [inline test/sharness/lib/0001-Generate-partial-JUnit-reports.patch]: Can we have it noop when JUnit is not needed? It is file IO, it can be slow.

File: test/sharness/lib/0001-Generate-partial-JUnit-reports.patch [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
++		echo "$*" >> .junit/tout
+ 		printf "%s" "$*"
+ 		tput sgr0
+ 		echo
+@@ -115,6 +117,8 @@ else
+ 	say_color() {
+ 		test -z "$1" && test -n "$quiet" && return
+ 		shift
++
++		echo "$*" >> .junit/tout
```
Comment (Kubuxu) [inline test/sharness/lib/0001-Generate-partial-JUnit-reports.patch]: Same in here.

File: test/sharness/lib/0001-Generate-partial-JUnit-reports.patch [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
++#!/usr/bin/env bash
+ #
+ # Copyright (c) 2011-2012 Mathias Lafeldt
+ # Copyright (c) 2005-2012 Git project
+@@ -106,6 +106,8 @@ if test -n "$color"; then
+ 			test -n "$quiet" && return;;
+ 		esac
+ 		shift
++
++		echo "$*" >> .junit/tout
```
Comment (Kubuxu) [inline test/sharness/lib/0001-Generate-partial-JUnit-reports.patch]: And here.

File: test/sharness/lib/0001-Generate-partial-JUnit-reports.patch [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
++	junit_testcase "$@"
+ }
+ 
+ # Public: Execute commands in debug mode.
+@@ -310,7 +359,7 @@ test_pause() {
+ test_eval_() {
+ 	# This is a separate function because some tests use
+ 	# "return" to end a test_expect_success block early.
+-	eval </dev/null >&3 2>&4 "$*"
++	eval </dev/null > >(tee -a .junit/tout >&3) 2> >(tee -a .junit/terr >&4) "$*"
```
[the remark this answers is not in the document]
Comment (Kubuxu) [reply test/sharness/lib/0001-Generate-partial-JUnit-reports.patch]: and in general.

**Comment by magik6k** [comment]
I've hacked together a pipeline to test how it looks / if it works, result here: https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fgo-ipfs/detail/PR-4530/10/tests

Since go-ipfs has few orders of magnitude more complex build system (sharness, plugins, integration testing in future, etc.), I may try to expand the pipeline I created to do the things we want it to do instead of trying to put that into the `golang` pipeline (so this PR would also replace #4430). Opinions?

**Comment by whyrusleeping** [comment]
@magik6k that looks really nice!

**Comment by Kubuxu** [comment]
You might want to disable colors, check the `runArgs` in Jenkinsfile.

File: ci/Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+def call(opts = []) {
+  def env = defVal(opts['env'], defaultEnv)
+  def test = defVal(opts['test'], defaultTest)
+  stage('tests') {
+    parallel(
+      linux: {
+        node(label: 'linux') {
+          ansiColor('xterm') {
+            def root = tool name: '1.9.2', type: 'go'
```
Comment (Kubuxu) [inline ci/Jenkinsfile]: We are currently running 1.10, do we have to duplicate this?

**Comment by Kubuxu** [comment]
Also if you take a look at: https://ci.ipfs.team/job/IPFS/job/go-ipfs/job/PR-4530/10/testReport/
You can see that sharness tests are split per file but all go tests are under the `(root)`. Can we have package `(sharness)` and go tests under their corresponding packages. IDK if that makes sense but I would like this UX to be nicer.

File: ci/Jenkinsfile (hunk 8 above)
Comment (magik6k) [reply ci/Jenkinsfile]: It was mostly copied from https://github.com/ipfs/jenkins-libs/blob/master/vars/golang.groovy, the jenkinsfile here in POC which I don't want merged (at least definitely not in this state)

**Comment by magik6k** [comment]
It's due to the way Jenkins parses the report file, here is what is generated:

https://ipfs.io/ipfs/QmYSVEtMSFzy5bh1fP8ynJnX6q3i8hYVKGuk4dtxV6gEWC/sharness.xml

In theory it's doable - generate a single, huge `<testsuite>` and put the test name into `classname` field of testcases. One problem is that you need to aggregate the results and doing it with a single `<testsiute>` is going to be far harder (at least in bash).

**Comment by Kubuxu** [comment]
You can always write a helper in Go.

File: ci/Jenkinsfile (hunk 8 above)
Comment (VictorBjelkholm) [reply ci/Jenkinsfile]: Yes, is absolutely POC and super messy atm. Look at the javascript one for a preview of final state: https://github.com/ipfs/jenkins-libs/blob/master/vars/javascript.groovy, notice the top variables defining platforms + versions to test with.

**Comment by magik6k** [comment]
Update:
* I cleaned up the Jenkins pipeline a lot
  * It doesn't test fuse currently, it tends to break workers
  * The current pipeline doesn't test fuse either so it's not a downgrade
  * Example build: https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fgo-ipfs/detail/PR-4530/146/pipeline
* Did the stuff @Kubuxu asked for - https://ci.ipfs.team/job/IPFS/job/go-ipfs/job/PR-4530/146/testReport/
* It's 3-4 mins faster than the current pipeline
* It's also building dists for multiple platforms: https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fgo-ipfs/detail/PR-4530/146/artifacts
* There are couple of failing tests -  https://ci.ipfs.team/blue/organizations/jenkins/IPFS%2Fgo-ipfs/detail/PR-4530/146/tests
  *  `windows / reprovide.TestReprovide` has a fix in go-libp2p-peer which isn't yet propagated
  * `linuxSharness / 2 - 'ipfs daemon --init'` - https://github.com/ipfs/go-ipfs/pull/4816
  * `macOS / util.TestManageInvalidNFds` is likely related to some worker configuration
  * `linuxSharness / docker` - Need to look at it more
  * other fails - need a look too

**Comment by Kubuxu** [comment]
@magik6k can we not increase the scope of this PR with benchmarking? Let's get it merged ASAP.

**Comment by magik6k** [comment]
Good point. Extracted the benchmark part into #4932

**Comment by Kubuxu** [comment]
@magik6k can we resolve the Jenkins being red? I am ok with even disabling OSX and Windows on Jenkins for time being until we can resolve it.

The JUnit reporting is important enough to have it merged before that can be resolved.

**Comment by magik6k** [comment]
It will be resolved by https://github.com/ipfs/jenkins/issues/107, which i hope will be resolved soon

**Comment by Kubuxu** [comment]
ipfs/jenkins#107 is resolved. RFM?
